src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,231 @@
+            }
+            else {
+                println("parent commit: null")
+            }
+
+            val diffs = df.scan(parentCommit, commit)
+
+            for (diff in diffs) {
+                // XXX(alex): does it happen in the wilds?
+                if (diff.changeType == DiffEntry.ChangeType.COPY) {
+                    continue
+                }
+
+                // File was renamed, tweak the files map.
+                if (diff.changeType == DiffEntry.ChangeType.RENAME) {
+                    println("old path: ${diff.getOldPath()}")
+                    println("new path: ${diff.getNewPath()}")
+                    files.set(diff.getOldPath(),
+                              files.remove(diff.getNewPath())!!)
+                    println("oopsy")
+                    //continue
+                }
+
+                // Get a edit list and traverse it backwards to avoid indices
+                // adjustment.
+                val editList = df.toFileHeader(diff).toEditList().asReversed()
+                for (edit in editList) {
+                    val delStart = edit.getBeginA()
+                    val delEnd = edit.getEndA()
+                    val delCount = edit.getLengthA()
+                    var insStart = edit.getBeginB()
+                    var insEnd = edit.getEndB()
+                    val insCount = edit.getLengthB()
+                    println("del ($delStart, $delEnd), ins ($insStart, $insEnd)");
+
+                    // Deletion case. Chaise down the deleted lines through the
+                    // history.
+                    if (delCount > 0) {
+                        val oldPath = diff.getOldPath()
+                        var tmpLines = ArrayList<RevCommitLine>(delCount)
+                        var idx = delStart;
+                        while (idx < delEnd) {
+                            tmpLines.add(RevCommitLine(commit, oldPath, idx));
+                            idx++
+                        }
+
+                        files.get(oldPath)!!.addAll(delCount, tmpLines)
+                    }
+
+                    // Insertion case. Report it.
+                    if (insCount > 0) {
+                        val newPath = diff.getNewPath()
+                        var newLines = files.get(newPath)!!
+                        val fileLoader = repo.open(diff.getNewId().toObjectId())
+                        val fileText = RawText(fileLoader.getBytes())
+
+                        var idx = insStart
+                        while (idx < insEnd) {
+                            val from = RevCommitLine(commit, newPath, idx)
+                            var to = newLines.get(idx)
+                            val cl = CodeLine(from, to, fileText.getString(idx))
+                            codeLines.add(cl)
+                            idx++
+                        }
+                        newLines.subList(insStart, insEnd).clear()
```

Comment:
Okay, at this point, it's too many comments inline for me to understand what the code does :). I will wait for another iteration, and then comment more.